Libraries/LibWeb/Fetch/Fetching/Fetching.cpp

```diff
@@ -2240,7 +2240,7 @@ WebIDL::ExceptionOr<GC::Ref<PendingResponse>> nonstandard_resource_loader_file_o
 
     auto request = fetch_params.request();
 
-    auto& page = Bindings::principal_host_defined_page(realm);
```

Comment:
This seems like an awkward footgun. Should we assert in principal_host_defined_page() that we're a principal realm? or is that the crash that we're fixing :thinking: . If `principal only` APIs should only be used by Principal realms perhaps we could encode that in the type? Or automatically grab the principal realm if the realm is-a shadow realm? I suspect the places that should really care are limited.

Reply:
Right, `principal_host_defined_XXX` asserts that it is given a principal realm, that's the crash. I don't think it is possible for us to encode that in the type - the only way that we know something is a principal realm is from the realm's host defined.

I'm not sure how many cases will actually need to access a page from the synthetic / shadow realm. I was also considering if we should just give a synthetic realm a pointer to a page as well, but I haven't quite figured out the extent of which we need to have page exposed given the limited set of APIs that a shadow realm has.

Reply:
Eh. All right. I'll put those thoughts back on the shelf until we have more test coverage for shadow realm exposed APIs actually using the realm